django/db/models/sql/compiler.py

```diff
@@ -1731,6 +1732,22 @@ def assemble_as_sql(self, fields, value_rows):
 
         return placeholder_rows, param_rows
 
+    def compile_where_clause_to_sql(self, expression):
+        if expression is None:
+            return "", []
+        if not isinstance(expression, (Q, F)):
+            raise ValueError(
+                "where_clause: The expression must be an instance of Q or F"
+            )
+        model = self.query.model
+        query = Query(model)
+        query.add_q(expression)
+        compiler = query.get_compiler(connection=connection)
+        node = query.where
+        sql, params = compiler.compile(node)
+
+        return sql, params
```

Comment:
It's probably worth making sure that `params` is a tuple here as per other methods:

```suggestion
        return sql, tuple(params)
```

Reply:
the returned parameters are not tuples but arrays. they will be used here: `param_rows += [params_on_conflict] `in the **compiler.py** file.